linera-faucet/server/src/lib.rs

```diff
@@ -182,72 +214,282 @@ where
+                        if i > 0
+                            && matches!(
+                                *exec_err,
+                                ExecutionError::BlockTooLarge
+                                    | ExecutionError::FeesExceedFunding { .. }
+                                    | ExecutionError::InsufficientBalance { .. }
+                                    | ExecutionError::MaximumFuelExceeded(_)
+                            ) =>
+                    {
+                        tracing::error!(%exec_err, "Execution of operation {i} failed; reducing batch size");
+                        self.config.max_batch_size = i as usize;
+                        // Put the valid requests back into the queue.
+                        let mut pending_requests = self.pending_requests.lock().await;
+                        for request in valid_requests.into_iter().rev() {
+                            pending_requests.push_front(request);
+                        }
+                        return Ok(()); // Don't return an error, so we retry.
+                    }
+                    chain_err => {
+                        return Err(
+                            ChainClientError::LocalNodeError(LocalNodeError::WorkerError(
+                                WorkerError::ChainError(chain_err.into()),
+                            ))
+                            .into(),
+                        )
+                    }
+                }
+            }
+            Err(err) => return Err(err.into()),
+            Ok(ClientOutcome::Committed(certificate)) => certificate,
+            Ok(ClientOutcome::WaitForTimeout(timeout)) => {
+                let error_msg = format!(
                     "This faucet is using a multi-owner chain and is not the leader right now. \
                     Try again at {}",
                     timeout.timestamp,
-                )));
+                );
+                for request in valid_requests {
+                    let _ = request.responder.send(Err(Error::new(error_msg.clone())));
+                }
+                return Ok(());
             }
         };
 
-        // Store the new mapping and save to disk
+        // Parse chain descriptions from the block's blobs
+        let blobs = certificate.block().body.blobs.iter().flatten();
+        let chain_descriptions = blobs
+            .map(|blob| bcs::from_bytes::<ChainDescription>(blob.bytes()))
+            .collect::<Result<Vec<ChainDescription>, _>>()?;
+
+        if chain_descriptions.len() != valid_requests.len() {
+            let error_msg = format!(
+                "Mismatch between operations ({}) and results ({})",
+                valid_requests.len(),
+                chain_descriptions.len()
+            );
+            for request in valid_requests {
+                let _ = request.responder.send(Err(Error::new(error_msg.clone())));
+            }
+            return Err(anyhow::anyhow!(error_msg));
```

Comment:
What if we had multiple requests to open the same chain - we'll probably end up opening only one (with the introduction of the recent feature that stores the owner->chain_id mapping) so this will trigger the error.

Reply:
No, we'd end up creating multiple chains, since the key is not in the map yet.

I don't think that's a problem, but we could deduplicate requests in each batch to address it. (Maybe in a separate PR?)